client/pkg/infra/infra_test.go

```diff
@@ -0,0 +1,172 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package infra_test
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/cosi-project/runtime/pkg/resource"
+	"github.com/cosi-project/runtime/pkg/resource/rtestutils"
+	"github.com/cosi-project/runtime/pkg/state"
+	"github.com/cosi-project/runtime/pkg/state/impl/inmem"
+	"github.com/cosi-project/runtime/pkg/state/impl/namespaced"
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zaptest"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/siderolabs/omni/client/api/omni/specs"
+	cloudspecs "github.com/siderolabs/omni/client/api/omni/specs/cloud"
+	"github.com/siderolabs/omni/client/pkg/infra"
+	"github.com/siderolabs/omni/client/pkg/infra/provision"
+	"github.com/siderolabs/omni/client/pkg/omni/resources/cloud"
+	"github.com/siderolabs/omni/client/pkg/omni/resources/omni"
+)
+
+type ms struct {
+	uuid string
+	id   string
+}
+
+type provisioner struct {
+	machines   map[resource.ID]ms
+	machinesMu sync.Mutex
+}
+
+// Provision implements provision.Provisioner interface.
+func (p *provisioner) Provision(_ context.Context, _ *zap.Logger, state *TestResource, request *cloud.MachineRequest) (provision.Result, error) {
+	p.machinesMu.Lock()
+	defer p.machinesMu.Unlock()
+
+	if p.machines == nil {
+		p.machines = map[string]ms{}
+	}
+
+	if state.TypedSpec().Value == nil {
+		state.TypedSpec().Value = &specs.MachineSpec{}
+	}
+
+	if state.TypedSpec().Value.Connected {
+		m := p.machines[request.Metadata().ID()]
+
+		return provision.Result{
+			UUID:      m.uuid,
+			MachineID: m.id,
+		}, nil
+	}
+
+	p.machines[request.Metadata().ID()] = ms{
+		uuid: uuid.New().String(),
+		id:   fmt.Sprintf("machine%d", len(p.machines)),
+	}
+
+	state.TypedSpec().Value.Connected = true
+
+	time.Sleep(time.Second)
```

Comment:
why do we need this? can we do it in another way?

Reply:
I verify that `MachineRequestStatus` enters `PROVISIONING` state. It requires some delay in the provisioner otherwise it completes too quickly and we don't see that state.

As another option I can make the test to unlock the processing manually after we get `PROVISIONING` phase reported.